src/cli.rs

```diff
@@ -0,0 +1,273 @@
+use crate::common::MetadataArgs;
+use bpaf::*;
+use std::{path::PathBuf, time::Duration};
+
+/// Arguments for typical querying commands - crates, publishers, json
+#[derive(Clone, Debug)]
+pub(crate) struct QueryCommandArgs {
+    pub cache_max_age: Duration,
+    pub diffable: bool,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum CliArgs {
+    Publishers {
+        args: QueryCommandArgs,
+        meta_args: MetadataArgs,
+    },
+    Crates {
+        args: QueryCommandArgs,
+        meta_args: MetadataArgs,
+    },
+    Json {
+        args: QueryCommandArgs,
+        meta_args: MetadataArgs,
+    },
+    JsonSchema {
+        print_schema: (),
+    },
+    Update {
+        cache_max_age: Duration,
+    },
+}
+
+pub(crate) fn args_parser() -> OptionParser<CliArgs> {
+    let diffable = short('d')
+        .long("diffable")
+        .help("Make output more friendly towards tools such as `diff`")
+        .switch();
+    let cache_max_age_parser = long("cache-max-age")
+        .help(
+            "\
+The cache will be considered valid while younger than specified.
+The format is a human readable duration such as `1w` or `1d 6h`.
+If not specified, the cache is considered valid for 48 hours.",
+        )
+        .argument("AGE")
+        .parse(|text| humantime::parse_duration(&text))
+        .fallback(Duration::from_secs(48 * 3600));
+    let cache_max_age = cache_max_age_parser.clone();
+    let args_parser = construct!(QueryCommandArgs {
+        cache_max_age,
+        diffable,
+    });
+
+    let all_features = long("all-features")
+        .help("Activate all available features")
+        .switch();
+    let no_default_features = long("no-default-features")
+        .help("Do not activate the `default` feature")
+        .switch();
+    let features = long("features")
+        .help("Space or comma separated list of features to activate")
+        .argument("FEATURES")
+        .optional();
+    let target = long("target")
+        .help("Only include dependencies matching the given target-triple")
+        .argument("TRIPLE")
+        .optional();
+    let manifest_path = long("manifest-path")
+        .help("Path to Cargo.toml")
+        .argument_os("PATH")
+        .map(|s| PathBuf::from(s))
+        .optional();
+
+    let metadata_args_parser = construct!(MetadataArgs {
+        all_features,
+        no_default_features,
+        features,
+        target,
+        manifest_path,
+    });
+
+    fn subcommand_with_common_args(
```

Comment:
Interesting approach. Is defining commands one by one without this helper that more verbose?

Reply:
`construct!` moves the `args` and `meta_args`, which requires explicitly cloning them for every subcommand. With two lines of `clone()` prepended to the definition of each subcommand it ended up being quite verbose. It was getting difficult to see what actually differs between subcommands.

If `construct!` implicitly cloned the parser, it wouldn't be quite as verbose and I probably wouldn't have bothered with the helper.

Reply:
Heh, though so, been on my TODO list for a while. Working on it right now actually :)

Reply:
Published 0.3.1, now you can do something like this

```rust
fn args() -> Parser<QueryCommandArgs> {
   ....
}

fn meta_args() -> Parser<MetadataArgs> {
   ....
}

...
  construct!(CliArgs::Publishers { args(), meta_args() }),
...
  construct!(CliArgs::Crates { args(), meta_args() }),
...
```